src/arcsf/data/data_module.py

```diff
@@ -107,31 +129,119 @@ def __init__(
-        """returns returns answer from a given row"""
-        return forget_row["answer"]
+    def get_answer(self, row):
+        """returns answer from a given row"""
+        return self.data[row]["answer"]
+
+    def qualitative_formatter(self, qa, _):
+        encoded_inp = self.tokenizer(
+            qa[0], return_tensors="pt", add_special_tokens=True
+        )
+        encoded_tar = self.tokenizer(
+            qa[1], return_tensors="pt", add_special_tokens=True
+        )
+        return (encoded_inp, encoded_tar)
+
+    def batch_formatter(
+        self, qa: tuple[str, str, int]
+    ) -> dict[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Formats the question-answer pair with padding and appropriately masked labels
+            allowing for batch computation.
+
+        Args:
+            qa : Tuple containing a question--answer pair
+
+        Returns:
+            formatted : formatted version of the input which can readily be passed to
+            model
+        """
+        question, _ = qa
+        encoded = self.tokenizer(
+            self.qa_formatter(qa),
+            max_length=self.max_length,
+            truncation=False,
+        )
+        num_question_tokens = len(self.tokenizer.tokenize(question))
+        pad_length = self.max_length - len(encoded.input_ids)
+        padded = encoded["input_ids"] + [self.tokenizer.eos_token_id] * pad_length
+        pad_attention_mask = encoded["attention_mask"] + [0] * pad_length
+        label = encoded.input_ids
+        if len(encoded.input_ids) == self.max_length:
+            label = encoded.input_ids
+        else:
+            label = (
+                encoded["input_ids"]
+                + [self.tokenizer.eos_token_id]
+                + [-100] * (pad_length - 1)
+            )
+
+        # change label to -100 for question tokens
+        for i in range(num_question_tokens):
+            label[i] = -100
+        return {
+            "input_ids": torch.tensor(padded).to(self.device),
+            "labels": torch.tensor(label).to(self.device),
+            "attention_mask": torch.tensor(pad_attention_mask).to(self.device),
+        }
+
+    def get_perturbed(
+        self, qa: tuple[str, str], row: int
+    ) -> list[dict[torch.Tensor, torch.Tensor, torch.Tensor]]:
+        """Returns batch-formatted version of the question answer pair along with
+            perturbed, erroneous question-answer pairs. For evaluation purposes.
+
+        Args:
+            qa : ground truth question answer tuple
+            row : question row ID for obtaining perturbed input
+
+        Returns:
+            formatted: batch formatted version of the perturbed inputs along with
+                formatted ground truth inputs
+        """
+        inp, _ = qa
+        author_q_n = row % self.data.TOFU_Q_PER_AUTHOR
+        perturbed_options = self.data[
+            row - author_q_n : row + self.data.TOFU_Q_PER_AUTHOR - author_q_n
+        ]["answer"]
```

Comment:
Exactly, that's what its doing. But I recognise now this doesn't generalise to a shuffled set. As you say, it would be much better to use the author index to find the relevant items. I will make this change.

Reply:
I think I've addressed this now

Reply:
Though, my new implementation runs self.data.filter() everytime __getitem__ is called. I will think of a way to calculate these at initialisation and store them?